From a69f86d9e022f8519bf89c3693e4244bb3c12513 Mon Sep 17 00:00:00 2001
From: John Chung <john.chung@mic.com.tw>
Date: Tue, 6 Apr 2021 15:44:17 +0800
Subject: [PATCH] Fix IPMI SEL entry in command  does not match IPMI SEL in
 file issue

Create ipmi_sel file if not exist
Add dbus match function for clear SEL feature
---
 .../xyz.openbmc_project.Logging.IPMI.service       |  1 +
 src/sel_logger.cpp                                 | 27 ++++++++++++++++++++++
 2 files changed, 28 insertions(+)
 mode change 100644 => 100755 service_files/xyz.openbmc_project.Logging.IPMI.service
 mode change 100644 => 100755 src/sel_logger.cpp

diff --git a/service_files/xyz.openbmc_project.Logging.IPMI.service b/service_files/xyz.openbmc_project.Logging.IPMI.service
old mode 100644
new mode 100755
index 9b8ae41..c75f60d
--- a/service_files/xyz.openbmc_project.Logging.IPMI.service
+++ b/service_files/xyz.openbmc_project.Logging.IPMI.service
@@ -3,6 +3,7 @@ Description=IPMI SEL Logging Service
 
 [Service]
 Restart=always
+ExecStartPre=/bin/bash -c 'if [ ! -f /var/lib/ipmi/ipmi_sel ]; then mkdir /var/lib/ipmi/; touch /var/lib/ipmi/ipmi_sel; fi'
 ExecStart=/usr/bin/sel-logger
 Type=simple
 
diff --git a/src/sel_logger.cpp b/src/sel_logger.cpp
old mode 100644
new mode 100755
index 7fe52fb..2050516
--- a/src/sel_logger.cpp
+++ b/src/sel_logger.cpp
@@ -28,6 +28,24 @@
 #include <sstream>
 #include <threshold_event_monitor.hpp>
 
+static bool clearSelEnable = false;
+
+inline static sdbusplus::bus::match::match startClearSelMonitor(
+    std::shared_ptr<sdbusplus::asio::connection> conn)
+{
+    auto clearSelMatcherCallback = [conn](sdbusplus::message::message &msg)
+    {
+            bool value;
+            msg.read(value);
+            clearSelEnable = value;
+    };
+    sdbusplus::bus::match::match clearSelMatcher(
+        static_cast<sdbusplus::bus::bus &>(*conn),
+        "type='signal',member='ClearSelEvent'",
+        std::move(clearSelMatcherCallback));
+    return clearSelMatcher;
+}
+
 struct DBusInternalError final : public sdbusplus::exception_t
 {
     const char *name() const noexcept override
@@ -107,6 +125,12 @@ static unsigned int getNewRecordId(void)
         recordId = selInvalidRecID;
     }
 
+    if(clearSelEnable)
+    {
+        recordId = selInvalidRecID;
+        clearSelEnable = false;
+    }
+
     if (++recordId >= selInvalidRecID)
     {
         recordId = 1;
@@ -200,6 +224,9 @@ int main(int argc, char *argv[])
         });
     ifaceAddSel->initialize();
 
+    sdbusplus::bus::match::match clearSelMonitor =
+        startClearSelMonitor(conn);
+
 #ifdef SEL_LOGGER_MONITOR_THRESHOLD_EVENTS
     sdbusplus::bus::match::match thresholdEventMonitor =
         startThresholdEventMonitor(conn);
-- 
2.7.4

